File: templates/Enterprise-Template/src/csharp/EnterpriseBotTemplate/Bot Framework/Dialogs/Main/MainResponses.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.IO;
using AdaptiveCards;
using $safeprojectname$.Dialogs.Main.Resources;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.TemplateManager;
using Microsoft.Bot.Schema;

namespace $safeprojectname$.Dialogs.Main
{
    public class MainResponses : TemplateManager
    {
        private static LanguageTemplateDictionary _responseTemplates = new LanguageTemplateDictionary
        {
            ["default"] = new TemplateIdMap
            {
                { ResponseIds.Cancelled,
                    (context, data) => 
                    MessageFactory.Text(
                        text: MainStrings.CANCELLED, 
                        ssml: MainStrings.CANCELLED, 
                        inputHint: InputHints.IgnoringInput)
                },
                { ResponseIds.Completed,
                    (context, data) => 
                    MessageFactory.Text(
                        text: MainStrings.COMPLETED, 
                        ssml: MainStrings.COMPLETED, 
                        inputHint: InputHints.IgnoringInput)
                },
                { ResponseIds.Confused,
                    (context, data) => 
                    MessageFactory.Text(
                        text: MainStrings.CONFUSED, 
                        ssml: MainStrings.CONFUSED, 
                        inputHint: InputHints.IgnoringInput)
                },
                { ResponseIds.Greeting,
                    (context, data) => 
                    MessageFactory.Text(
                        text: MainStrings.GREETING, 
                        ssml: MainStrings.GREETING, 
                        inputHint: InputHints.IgnoringInput)
                },
                { ResponseIds.Help, (context, data) => BuildHelpCard(context, data) },
                { ResponseIds.Intro, (context, data) => BuildIntroCard(context, data) },
            }
        };

        public MainResponses()
        {
            Register(new DictionaryRenderer(_responseTemplates));
        }

        public static IMessageActivity BuildIntroCard(ITurnContext turnContext, dynamic data)
        {
            var introCard = File.ReadAllText(@".\Dialogs\Main\Resources\Intro.json");
            var card = AdaptiveCard.FromJson(introCard).Card;
            var attachment = new Attachment(AdaptiveCard.ContentType, content: card);

            var response = MessageFactory.Attachment(attachment, ssml: card.Speak, inputHint: InputHints.AcceptingInput);
            response.SuggestedActions = new SuggestedActions()
            {
                Actions = new List<CardAction>()
                {
                    new CardAction(type: ActionTypes.ImBack, title: "Test LUIS", value: "Hi there!"),
                    new CardAction(type: ActionTypes.ImBack, title: "Test QnAMaker", value: "Why did Microsoft develop the Bot Framework?"),
                    new CardAction(type: ActionTypes.OpenUrl, title: "Learn More", value: "https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-enterprise-template-overview?view=azure-bot-service-4.0")
                }
            };

            return response;
        }

        public static IMessageActivity BuildHelpCard(ITurnContext turnContext, dynamic data)
        {
            var attachment = new HeroCard()
            {
                Title = MainStrings.HELP_TITLE,
                Text = MainStrings.HELP_TEXT,
                Buttons = new List<CardAction>()
                {
                    new CardAction(type: ActionTypes.ImBack, title: "Test LUIS", value: "Hello"),
                    new CardAction(type: ActionTypes.ImBack, title: "Test QnAMaker", value: "What is the sdk v4?"),
                    new CardAction(type: ActionTypes.OpenUrl, title: "Learn More", value: "https://docs.microsoft.com/en-us/azure/bot-service/?view=azure-bot-service-4.0"),
                },
            }.ToAttachment();

            var response = MessageFactory.Attachment(attachment, ssml: MainStrings.HELP_TEXT, inputHint: InputHints.AcceptingInput);
            return response;
        }

        public class ResponseIds
        {
            // Constants
            public const string Cancelled = "cancelled";
            public const string Completed = "completed";
            public const string Confused = "confused";
            public const string Greeting = "greeting";
            public const string Help = "help";
            public const string Intro = "intro";
        }
    }
}

File: templates/Enterprise-Template/src/typescript/src/dialogs/main/mainResponses.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { DictionaryRenderer, LanguageTemplateDictionary, TemplateFunction } from "../templateManager/dictionaryRenderer";
import { TemplateManager } from "../templateManager/templateManager";
import * as i18n from "i18n";
import { ActionTypes, Activity, CardFactory, TurnContext, MessageFactory, InputHints } from "botbuilder";
import { ActivityExtensions } from "../../extensions/activityExtensions";
const introCard = require("./resources/Intro.json");

export class MainResponses extends TemplateManager {
    
    // Fields
    public static ResponseIds = {
        Cancelled: "cancelled",
        Completed: "completed",
        Confused: "confused",
        Greeting:  "greeting",
        Help:  "help",
        Intro:  "intro",
    }
    private static readonly _responseTemplates: LanguageTemplateDictionary = new Map([
        ["default", new Map([
            [MainResponses.ResponseIds.Cancelled, MainResponses.fromResources("main.cancelled")],
            [MainResponses.ResponseIds.Completed, MainResponses.fromResources("main.completed")],
            [MainResponses.ResponseIds.Confused, MainResponses.fromResources("main.confused")],
            [MainResponses.ResponseIds.Greeting, MainResponses.fromResources("main.greeting")],
            [MainResponses.ResponseIds.Help, (context: TurnContext, data: any) => MainResponses.buildHelpCard(context, data)],
            [MainResponses.ResponseIds.Intro, (context: TurnContext, data: any) => MainResponses.buildIntroCard(context, data)]
        ])]
    ]);
 
    constructor() {
        super();
        this.register(new DictionaryRenderer(MainResponses._responseTemplates));
    }

    private static fromResources(name: string): TemplateFunction {
        return (context: TurnContext, data: any) => Promise.resolve(i18n.__(name));
    }

    public static buildIntroCard(turnContext: TurnContext, data: any): Promise<Activity> {
        const introPath = i18n.__("main.introPath");
        const introCard = require(introPath);
        const attachment = CardFactory.adaptiveCard(introCard);
        const response = MessageFactory.attachment(attachment, '', attachment.content.speak, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
                {
                    title: 'Test LUIS',
                    type: ActionTypes.ImBack,
                    value: 'Talk to a human',  
                },
                {
                    title: 'Test QnA Maker',
                    type: ActionTypes.ImBack,
                    value: 'What is the Enterprise Bot Template?',  
                },
                {
                    title: 'Learn more',
                    type: ActionTypes.OpenUrl,
                    value: 'https://docs.microsoft.com/en-us/azure/bot-service/?view=azure-bot-service-4.0',  
                },
            ],
            to:[],
        };

        return Promise.resolve(response as Activity);      
    }

    public static async buildHelpCard(turnContext: TurnContext, data: any): Promise<Activity> {
        const title = i18n.__("main.helpTitle");
        const text = i18n.__("main.helpText");
        var attachment = CardFactory.heroCard(title,text);
        var response = MessageFactory.attachment(attachment,text,InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
            {
                title: 'Test LUIS',
                type: ActionTypes.ImBack,
                value: 'Talk to a human',  
            },
            {
                title: 'Test QnA Maker',
                type: ActionTypes.ImBack,
                value: 'What is the Enterprise Bot Template?',  
            },
            {
                title: 'Learn more',
                type: ActionTypes.OpenUrl,
                value: 'https://docs.microsoft.com/en-us/azure/bot-service/?view=azure-bot-service-4.0',  
            },
            ],
            to:[],
        };

        return Promise.resolve(response as Activity);    
    }        
}